### PR TITLE
Track source of vars

### PR DESCRIPTION
**Pull request by agaffney**
##### SUMMARY
Track the source of variables when calling `get_vars()` in `VarsManager`

This is useful for debugging where ansible is getting a particular variable and for doing things like throwing deprecation warnings when using variables from certain sources (such as top-level facts).

This implementation cheats a bit by tracking which precedence level (and which file source, if available) a var comes from as the varset is built in `get_vars()`, rather than tracking each var from its original source. While creating a variable proxy/wrapper class and using it for each var would be the "better" approach, it's not practical to do so for simple variable types such as `bool` and `int` without breaking a lot of things or requiring a lot of hacks, mostly due to the JSON encoder.

This PR is the spiritual successor for #41811

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
VarsManager

##### ADDITIONAL INFORMATION
N/A

### CODE REVIEW

File: lib/ansible/vars/manager.py [hunk 1]
```diff
@@ -661,3 +674,20 @@ def set_host_variable(self, host, varname, value):
             self._vars_cache[host] = combine_vars(self._vars_cache[host], {varname: value})
         else:
             self._vars_cache[host][varname] = value
+
+
+class VarsWithSources(dict):
```
Comment (abadger) [inline lib/ansible/vars/manager.py]: Although you have to implement a little more, use MutableMapping as the base class instead of dict.  (There's supposedly corner cases with dict)

Comment (agaffney) [reply lib/ansible/vars/manager.py]: I know there are corner cases when subclassing `dict`, but I don't think any of them apply here. This passes all CI tests.

File: lib/ansible/vars/manager.py [hunk 2]
```diff
@@ -661,3 +674,20 @@ def set_host_variable(self, host, varname, value):
...
+    '''
+    Dict-like class for vars that also provides source information for each var
+    '''
+    def __init__(self, data, sources):
```
Comment (abadger) [inline lib/ansible/vars/manager.py]: This `__init__` signature is not dict-compatible.  I know this has come up before and there was a workaround but I'm not remembering it right this instant (and the problem might have been easier to resolve, then).

The workarounds I can think of right now:

* Set sources via a method.  It's not conceptually the best as sources is really something that is set once when the object is instantiated and then should not be set again.
* Use a special name with a keyword arg that minimizes the chance of conflicts.    Also not the best as there's always the chance that someone will add that as a conflicting key somewhere down the line.  If you go this route, be aware that Python2 is less flexible than Python3:
``` python
# Works on Python3 but fails on Python2:
def mine(*args, __ansible_DEADBEEF64_source='', **kwargs):    
    print(args)    
    print(__ansible_DEADBEEF64_source)    
    print(kwargs)    
    
mine()    
mine(1, 2)    
mine(1, 2, data='test')    
mine(1, 2, data='test', __ansible_DEADBEEF64_source='from user')    
mine(data='test', __ansible_DEADBEEF64_source='from user') 
```

You have to include the special variable in kwargs and then pop it to be compatible with Python2.

File: lib/ansible/vars/manager.py [hunk 3]
```diff
@@ -179,6 +179,14 @@ def get_vars(self, play=None, host=None, task=None, include_hostvars=True, inclu
             _hosts_all=_hosts_all,
         )
 
+        _vars_sources = dict()
```
Comment (abadger) [inline lib/ansible/vars/manager.py]: Note that we've moved to asking people to use literal notation for empty dict, list, etc a few years ago.

Comment (abadger) [reply lib/ansible/vars/manager.py]:
```suggestion
        _vars_sources = {}
```

File: lib/ansible/vars/manager.py [hunk 4]
```diff
@@ -661,3 +674,20 @@ def set_host_variable(self, host, varname, value):
...
+    Dict-like class for vars that also provides source information for each var
+    '''
+    def __init__(self, data, sources):
+        super(VarsWithSources, self).__init__(data)
+        self.sources = sources
+
+    def get_source(self, key):
+        return self.sources.get(key, None)
+
+    def __getitem__(self, key):
```
Comment (sivel) [inline lib/ansible/vars/manager.py:709]: This is actually one reason you want to use `MutableMapping` over subclassing `dict`.  `.get()` does not use `.__getitem__()` in `dict`. Both methods duplicate code for performance:

```pycon
>>> class VarsWithSources(dict):
...     def __getitem__(self, key):
...         val = super(VarsWithSources, self).__getitem__(key)
...         print(('__getitem__', val))
...         return val
...
>>> v = VarsWithSources()
>>> v['foo'] = 'bar'
>>> v['foo']
('__getitem__', 'bar')
'bar'
>>> v.get('foo')
'bar'
```

Notice that `.get()` doesn't cause the print to happen.  As such, your implementation as is will only work with some accesses of keys, and not others.

File: lib/ansible/vars/manager.py (hunk 3 above)
Comment (agaffney) [reply lib/ansible/vars/manager.py]: Meaning `_vars_sources = {}`?

File: lib/ansible/vars/manager.py (hunk 2 above)
Comment (agaffney) [reply lib/ansible/vars/manager.py]: The `__init__` signature doesn't need to be dict compatible, because it will never be unknowingly used in place of a dict anywhere that a new object would be created.

File: lib/ansible/vars/manager.py (hunk 3 above)
Comment (agaffney) [reply lib/ansible/vars/manager.py]: Ah, I see you put that in a suggestion. That suggested change doesn't show up in the conversation view on the issue...

**Comment by abadger** [comment]
Does this implementation only tag the toplevel vars with source? If so, you should add a comment in a few places (I don't see an obvious single place where the comment should live) that notes that is the case.

**Comment by agaffney** [comment]
Yes, this only tracks top-level vars. I'm not even sure how you would track non-top-level vars to make it meaningful to add that comment.

**Comment by abadger** [comment]
<nod>  But the fact that it is very hard to track non-top-level vars is not a reason not to comment that only toplevel vars are tracked.

File: lib/ansible/vars/manager.py (hunk 2 above)
Comment (abadger) [reply lib/ansible/vars/manager.py]: *Right now.*  But that isn't something that should be depended upon in general and especially if you are creating a subclass of a builtin type.

I did recall the non-hacky way to fix this, though.  Change the constructor to be dict-compatible and then create an alternate constructor using classmethod to do what you want.  Then use the classmethod in your code.  Looks like this:

``` python
class VarsWithSources(MutableMapping):
    def __init__(self, data):
        self.sources = "<Source unset>"
        super(self, VarsWithSources).__init__(data)

    @classmethod
    def new_vars_with_source(cls, data, sources):
        new_vars_with_source = cls(data)
        new_vars_with_source.sources = sources
        return new_vars_with_source
```

Comment (abadger) [reply lib/ansible/vars/manager.py]: And then where you call it you do:
```python
VarsWithSources.new_vars_with_source(all_vars, _vars_sources)
```

File: lib/ansible/vars/manager.py [hunk 5]
```diff
@@ -661,3 +677,51 @@ def set_host_variable(self, host, varname, value):
...
+    def new_vars_with_sources(cls, data, sources):
+        ''' Static class method to instantiate class with sources '''
+        v = cls(data)
+        v.sources = sources
+        return v
+
+    def get_source(self, key):
+        return self.sources.get(key, None)
+
+    def __getitem__(self, key):
```
Comment (abadger) [inline lib/ansible/vars/manager.py]:
```suggestion
    def __getitem__(self, key):
        # See notes in VarsWithSources' docstring for caveats and limitations
```

File: lib/ansible/vars/manager.py [hunk 6]
```diff
@@ -661,3 +677,51 @@ def set_host_variable(self, host, varname, value):
...
+        val = self.data[key]
```
Comment (abadger) [inline lib/ansible/vars/manager.py:710]:
```suggestion
        val = self.data[key]
        
        # See notes in VarsWithSources' docstring for caveats and limitations of the source tracking
```

File: lib/ansible/vars/manager.py [hunk 7]
```diff
@@ -179,6 +179,16 @@ def get_vars(self, play=None, host=None, task=None, include_hostvars=True, inclu
             _hosts_all=_hosts_all,
         )
 
+        _vars_sources = {}
+
+        def _combine_and_track(data, new_data, source):
+            ''' Wrapper function to update var sources dict and call combine_vars() '''
```
Comment (abadger) [inline lib/ansible/vars/manager.py]:
```suggestion
            '''
            Wrapper function to update var sources dict and call combine_vars()
            
            See notes in VarsWithSources' docstring for caveats and limitations of the source tracking
            '''
```

File: lib/ansible/vars/manager.py [hunk 8]
```diff
@@ -661,3 +677,51 @@ def set_host_variable(self, host, varname, value):
...
+    when a particular var is accessed.
+    '''
+    def __init__(self, *args, **kwargs):
+        ''' Dict-compatible constructor '''
+        self.data = dict(*args, **kwargs)
+        self.sources = {}
+
+    @classmethod
+    def new_vars_with_sources(cls, data, sources):
+        ''' Static class method to instantiate class with sources '''
```
Comment (abadger) [inline lib/ansible/vars/manager.py]: Cosmetic:

```suggestion
        ''' Alternate constructor to instantiate class with sources '''
```

A bit more informative about what the method is intended to be used for.

**Comment by abadger** [comment]
Added a couple of suggestions for comments pointing to the primary documentation of how vars source tracking works.

Suggestion for the docstring of the new classmethod that might make it clearer to someone coming fresh to the code.

Looks good to me after the first two are merged.  The docstring change is at your option.

**Comment by abadger** [comment]
rebuild_merge

**Comment by abadger** [comment]
Failing tests are unrelated to this PR.  Merging.
